eras/babbage/impl/src/Cardano/Ledger/Babbage/Scripts.hs

```diff
@@ -0,0 +1,110 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Figure 3: Functions related to scripts
+--   Babbage Specification
+module Cardano.Ledger.Babbage.Scripts where
+
+import Cardano.Ledger.Alonzo.Data (BinaryData, dataToBinaryData)
+import Cardano.Ledger.Alonzo.Language (Language (..))
+import Cardano.Ledger.Alonzo.PlutusScriptApi (language)
+import Cardano.Ledger.Alonzo.Scripts (Script)
+import Cardano.Ledger.Alonzo.Tx (ScriptPurpose (..), txdats')
+import Cardano.Ledger.Alonzo.TxWitness (TxWitness, txscripts', unTxDats)
+import Cardano.Ledger.Babbage.TxBody
+  ( Datum (..),
+    TxBody (..),
+    TxOut (..),
+    referenceInputs',
+    spendInputs',
+  )
+import qualified Cardano.Ledger.Core as Core
+import Cardano.Ledger.Era (Crypto, Era, ValidateScript (hashScript, isNativeScript))
+import Cardano.Ledger.Shelley.Scripts (ScriptHash (..))
+import Cardano.Ledger.Shelley.UTxO (UTxO (..))
+import Cardano.Ledger.TxIn (TxIn)
+import qualified Data.Compact.SplitMap as SplitMap
+import qualified Data.Map as Map
+import Data.Set (Set)
+import qualified Data.Set as Set
+import GHC.Records (HasField (..))
+
+-- ====================================================================
+
+hasTxIn :: ScriptPurpose crypto -> Maybe (TxIn crypto)
+hasTxIn (Spending txin) = Just txin
+-- Only the Spending ScriptPurpose contains TxIn
+hasTxIn (Minting _policyid) = Nothing
+hasTxIn (Rewarding _rewaccnt) = Nothing
+hasTxIn (Certifying _dcert) = Nothing
+
+getDatum ::
+  ( Era era,
+    Core.TxOut era ~ TxOut era,
+    Core.Witnesses era ~ TxWitness era
+  ) =>
+  Core.Tx era ->
+  UTxO era ->
+  ScriptPurpose (Crypto era) ->
+  [BinaryData era]
+getDatum tx (UTxO m) sp = case hasTxIn sp of
+  Nothing -> []
+  Just txin ->
+    case SplitMap.lookup txin m of
+      Nothing -> []
+      Just (TxOut _ _ datum _refScript) ->
+        case datum of
+          NoDatum -> []
+          Datum d -> [d]
+          DatumHash hash ->
+            case Map.lookup hash (unTxDats $ txdats' (getField @"wits" tx)) of
+              Nothing -> []
+              Just dat -> [dataToBinaryData dat]
```

Comment:
This is what the Spec says, It returns a list.

Reply:
@TimSheard I think the spec should be fixed, but @WhatisRT said that it would be inconvenient, but we can use `Maybe` in the implementation instead:

> I think what you should really do instead is to have a Maybe in the implementation, and keep it in the spec as it is

Reply:
I am not so sure.  In the rules this is used to build a list of BinaryData which is then concateted with other lists, and then serialized to pass to the TxInfo. Something like this
serialize (getDatum x ++ getAnotherKindOfData y ++ getYetAnother z)

Reply:
Yeah as far as the spec is concerned, I understand that it can be more convenient to return a list, but then in our implementation it should:
```
serialize (maybeToList (getDatum x) ++ getAnotherKindOfData y ++ getYetAnother z)
```